pykokkos/lib/ufuncs.py

```diff
@@ -798,4 +798,538 @@ def logaddexp(viewA, viewB):
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+            Input view.
+    viewB : pykokkos view
+            Input view.
+
+    Returns
+    -------
+    out : pykokkos view
+           Output view.
+
+    """
+    if str(viewA.dtype) == "DataType.double" and str(viewB.dtype) == "DataType.double":
+        out = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            logaddexp2_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+
+    elif str(viewA.dtype) == "DataType.float" and str(viewB.dtype) == "DataType.float":
+        out = pk.View([viewA.shape[0]], pk.float)
+        pk.parallel_for(
+            viewA.shape[0],
+            logaddexp2_impl_1d_float,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+
+@pk.workunit
+def floor_divide_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = viewA[tid] // viewB[tid]
+
+@pk.workunit
+def floor_divide_impl_1d_float(tid: int, viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float], out: pk.View1D[pk.float]):
+    out[tid] = viewA[tid] // viewB[tid]
+
+
+def floor_divide(viewA, viewB):
+    """
+    Divides positionally corresponding elements
+    of viewA with elements of viewB and floors the result
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+            Input view.
+    viewB : pykokkos view
+            Input view.
+
+    Returns
+    -------
+    out : pykokkos view
+           Output view.
+
+    """
+    if str(viewA.dtype) == "DataType.double" and str(viewB.dtype) == "DataType.double":
+        out = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            floor_divide_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+
+    elif str(viewA.dtype) == "DataType.float" and str(viewB.dtype) == "DataType.float":
+        out = pk.View([viewA.shape[0]], pk.float)
+        pk.parallel_for(
+            viewA.shape[0],
+            floor_divide_impl_1d_float,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+
+@pk.workunit
+def sin_impl_1d_double(tid: int, view: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = sin(view[tid])
+
+@pk.workunit
+def sin_impl_1d_float(tid: int, view: pk.View1D[pk.float], out: pk.View1D[pk.float]):
+    out[tid] = sin(view[tid])
+
+def sin(view):
+    """
+    Element-wise trignometric sine of the view
+
+    Parameters
+    ----------
+    view : pykokkos view
+           Input view.
+
+    Returns
+    -------
+    out : pykokkos view
+           Output view.
+
+    """
+    if str(view.dtype) == "DataType.double":
```

Comment:
why do we need to convert to string?

Reply:
Good question--we've been doing this because we don't really have proper type object/type hierarchy implemented yet--you can see that we just have some "skeletons" in https://github.com/kokkos/pykokkos/blob/develop/pykokkos/interface/data_types.py. 

We probably eventually want something like [`np.can_cast`](https://numpy.org/doc/stable/reference/generated/numpy.can_cast.html) for casting-compatible hierarchy assessment.

I think Nader and I both liked the type promotion lattice that Google uses for Jax: https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=types (and they are also concerned with the array API conformance I think).